### PR TITLE
Make MemoryImage::new() failable + bound check

### LINKED ISSUE #531: [Bug] index out of bound  panic while creating a MemoryImage from a riscV program
# index out of bound  panic while creating a MemoryImage from a riscV program

We (@fuzzinglabs & @lambdaclass) found an index out of bound issue in risc0 while reading a RISC Zero Program and creating a MemoryImage.

## Expected Behavior

It should not crash.

## Your Environment

-  e862e0a4d940dce9618f673b6503751b1bc20608
-  cargo 1.70.0-nightly (15d090969 2023-03-21)
-  Ubuntu 22.04

## Steps to reproduce

Download: 

Testing program:
``` rust
use risc0_zkvm::{MemoryImage, Program, MEM_SIZE};

fn index_oob_risc0_program() {
    let data = b"\x7f\x45\x4c\x46\x01\x01\x01\x01\x01\x01\xff\xff\x00\x00\x00\x00\x02\x00\xf3\x00\x00\x00\x00\x00\x00\x00\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x20\x00\x08\x00\x00\x00\x96\x96\x00\x94\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\xff\x00\x00\x94\x00\x00\x00\xff\xf6\x12\xa9\x00\x00\x00\x00\x00\x00\xfe\x00\x00\x00\x00\x00\x0a\x9a\x38\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x96\x4c\x46\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x0a\x9d\xd8\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x40\x1e\x00\x00\x46\x4c\x00\x00\x00\x00\x00\x02\x00\x40\x00\x01\x01\x01\x00\x04\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x05\x00\x00\x07\x78\xc1\x0a\x00\x00\xba\x00\x00\x00\x00\xe3\x04\x00\x00\x31\x35\x32\x37\x38\x31\x46\x01\x01\x01\x01\x01\x01\xff\xff\x00\x00\x00\x00\x02\x00\xe5\x00\x00\x00\x00\x96\x96\x00\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\xff\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x0b\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x06\x2e\xac\x00\x00\x00\x00\x00\x00\x0a\xce\x58\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x00\x00\x00\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x02\x40\x1e\x1e\x00\x00\x00\x00\x00\x00\x00\x00\x02\x00\x40\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x05\x00\x00\x07\x00\xba\xe8\xad\x0a\x00\xe3\x04\x00\x00\x00\x00\x12\x00\x00\x00\x00\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x05\x00\x00\x00\x01\x01\x01\x50\xcf\x0a\x00\x01\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x00\x04\x01\x01\x01\x01\x01\x01\x01\x00\x00\x31\x35\x31\x35\x32\x37\x38\x31\x30\x34\x02\x00\x00\x00\x00\x00\x00\x00\x00\x05\x00\x00\x05\x00\x00\x00\x01\x01\x01\x01\x01\x01\x01\x00\x00\x00\x00\x00\x00\x07\x00\x00\x00\xff\xff\xff\xff\x00\x00\x00\x00\xff\x04\x92\x01\x01\x01\x01\x01\x01\xa2\xf8\x00\x20\x00\x00\x00\x00\xff\x00\x40\x00\x04\x00\x00\x00\x38\x00\x00\x00\x00\x00\x00\x00\x02\x00\x0a\x40\x40\x00\x1a\x00\x19\x00";
    const PAGE_SIZE: u32 = 1024;
    let _ = match Program::load_elf(data, MEM_SIZE as u32) {
        Ok(prog) => {
            MemoryImage::new(&prog, PAGE_SIZE);
            true
        }
        Err(_) => false,
    };
}
```

## Root cause

```
thread 'main' panicked at 'index out of bounds: the len is 268435456 but the index is 4278190080', /home/nabih/risc0_fuzzing/risc0/risc0/zkvm/src/binfmt/image.rs:132:17
```

Related code: https://github.com/risc0/risc0/blob/e862e0a4d940dce9618f673b6503751b1bc20608/risc0/zkvm/src/binfmt/image.rs#L130

## Detailed behavior
``` sh
thread 'main' panicked at 'index out of bounds: the len is 268435456 but the index is 4278190080', /home/nabih/risc0_fuzzing/risc0/risc0/zkvm/src/binfmt/image.rs:130:17
stack backtrace:
   0: rust_begin_unwind
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/core/src/panicking.rs:65:14
   2: core::panicking::panic_bounds_check
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/core/src/panicking.rs:151:5
   3: <usize as core::slice::index::SliceIndex<[T]>>::index_mut
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/core/src/slice/index.rs:259:14
   4: core::slice::index::<impl core::ops::index::IndexMut<I> for [T]>::index_mut
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/alloc/src/vec/mod.rs:2750:9
   5: <alloc::vec::Vec<T,A> as core::ops::index::IndexMut<I>>::index_mut
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/alloc/src/vec/mod.rs:2750:9
   6: risc0_zkvm::binfmt::image::MemoryImage::new
             at /home/nabih/risc0_fuzzing/risc0/risc0/zkvm/src/binfmt/image.rs:130:17
   7: load_elf::main::{{closure}}
             at ./src/main.rs:11:21
   8: honggfuzz::fuzz
             at /home/nabih/.cargo/registry/src/github.com-1ecc6299db9ec823/honggfuzz-0.5.55/src/lib.rs:317:5
   9: load_elf::main
             at ./src/main.rs:6:9
  10: core::ops::function::FnOnce::call_once
             at /rustc/b8b5caee04116c7383eb1c6470fcf15c437a60d4/library/core/src/ops/function.rs:251:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by mothran**
Fixes #531

### CODE REVIEW

File: risc0/zkvm/src/binfmt/image.rs [hunk 1]
```diff
@@ -119,23 +120,24 @@ impl MemoryImage {
...
         // Load the ELF into the memory image.
         for (addr, data) in program.image.iter() {
             let addr = *addr as usize;
             let bytes = data.to_le_bytes();
-            for i in 0..WORD_SIZE {
-                buf[addr + i] = bytes[i];
+            if (WORD_SIZE + addr) >= buf.len() {
+                bail!("Invalid Elf Program, address outside MEM_SIZE");
             }
+            buf[addr..(WORD_SIZE + addr)].copy_from_slice(&bytes[..WORD_SIZE]);
```
Comment (flaub) [inline risc0/zkvm/src/binfmt/image.rs:133]:
```suggestion
            buf.get(addr..(WORD_SIZE + addr))?.copy_from_slice(&bytes[..WORD_SIZE]);
```

Comment (mothran) [reply risc0/zkvm/src/binfmt/image.rs:133]: good call, we could drop the if + bail! then right?

Comment (mothran) [reply risc0/zkvm/src/binfmt/image.rs:133]: err I guess that would remove the friendly Error string

Comment (mothran) [reply risc0/zkvm/src/binfmt/image.rs]: Combined the two here: https://github.com/risc0/risc0/pull/533/commits/d610dc2fcf909398d3db07bfa3e98d5714615d80

**Review by flaub — APPROVED** [review]

**Pull request merged**